docs/writing_a_builder.md

```diff
@@ -31,6 +36,43 @@ will be passed as a primary input, and each build step may produce two assets.
 For the primary input `some_file.txt` the allowed outputs are
 `some_file.txt.foo` and `some_file.txt.bar`.
 
+### Capture groups
+
+Builders can declare more complex inputs and outputs by using a capture group
+in their build input. Capture groups allow builders to move files across
+directories. For instance, consider a builder that generates Dart code for
+[Protocol Buffers][protobuf]. Let's assume that proto definitions are stored in
+a top-level `proto/` folder, and that generated files should go to
+`lib/src/proto/`. This cannot be expressed with simple build extensions that
+may replace a suffix in the asset's path only.
+Using `{'proto/{{}}.proto': ['lib/src/proto/{{}}.dart']}` as a build extension
+lets the builder read files in `proto/` and emit Dart files in the desired 
+location. Here, the __`{{}}`__ is called a _capture group_. Capture groups have
+the following noteworthy properties:
+
+- Capture groups match at least one character in the input path, but may match
+  arbitrarily many characters.
+- When the input uses a capture group, every output must reference that capture
+  group as well.
+- It is not currently allowed to declare multiple capture groups in one input
+  extension.
+- Capture groups match as many characters as possible. In the proto example,
+  the asset `proto/nested/proto/test.proto` would match with `{{}}` capturing
+  `nested/proto/test`. The shorter suffix match with `{{}}` capturing just
+  `test` is not considered.
```

Comment:
Can we add a sentence with something like. "If the `{{}}` group is at the start of the input it will match the full path from the root of the package. To match only file names start the input expression with `/{{}}`.

Reply:
Because of the eager matching, this would actually match the entire path after the first `/` today.

Reply:
Oh right.

Hmm, does that mean this implementation does not cover this use case?
https://github.com/dart-lang/build/issues/1689#issuecomment-728194831

Reply:
Good point, I do think we need to think about how to enable that case. The current design works well for moving files from one top level directory to another, but not for moving files more locally as in this example.

Reply:
I think if we had a way for you to configure whether the match should match directory separators, that could enable this?

Should we have another type of capture group possibly? Or can we think of a better way to configure the existing type?

Reply:
One option could be just allowing multiple capture groups:

```yaml
build_extensions:
  '{{0}}/{{1}}.graphql':
    - '{{0}}/graphql/{{1}}.data.gql.dart'
```

Reply:
I think I prefer multiple capture groups if we can support them on all build implementations. They're more powerful and efficient than lazily-matched groups. Should I do that in this PR or open another one?

-----

Even if we allowed a lazy capture group, `\/(.+?)\.graphql$` would still start matching after the first `/`:

```dart
void main() {
  final regex = RegExp(r'\/(.+?)\.graphql$');
  
  // only finds a single match, with the group capturing `foo/bar/baz `
  regex.allMatches('lib/foo/bar/baz.graphql').forEach((match) {
    print(match.group(1));
  });
}
```

I don't know if there's a way to make a group capture the least amount of chars on the right, I'm not that good at regex :D We could implicitly start the regex with `.*` if there's no `^` in the input, but then we'd have problems with capture groups at the beginning.

Reply:
I think I prefer the multiple capture groups approach as well.

For bazel we can't use regex, and we also can't use recursion, which makes things a bit interesting. But I think it should be possible to implement still.

We almost certainly could support some hard coded number of capture groups in any case, which would be sufficient.

So I think just supporting multiple (named) capture groups is the way to go. I think leaving unnamed capture groups is still also fine.

Reply:
I've added support for multiple capture groups in 2349c1f7e516da07291374d2f2dba9eb4ae86458. I also added another subsection showing how to use them for the "generate in relative subdirectory" usecase.

Reply:
thanks, this LGTM 👍